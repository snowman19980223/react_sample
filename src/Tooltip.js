import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import {mapWhitespaceProps} from './props'

const alignmentDirections = ['ne', 'se', 'nw', 'sw']

export default function Tooltip(props) {
  const {children, direction, text, noDelay, align, wrap, className} = mapWhitespaceProps(props)
  return (
    <span
      aria-label={text}
      className={classnames(
        className,
        'tooltipped',
        `tooltipped-${direction}`,
        align && `tooltipped-align-${align}-2`,
        noDelay && 'tooltipped-no-delay',
        wrap && 'tooltipped-multiline'
      )}
    >
      {children}
    </span>
  )
}

Tooltip.directions = ['n', 'ne', 'e', 'se', 's', 'sw', 'w', 'nw']

Tooltip.defaultProps = {
  direction: 'n'
}

Tooltip.propTypes = {
  align: PropTypes.oneOf(alignmentDirections),
  children: PropTypes.node,
  direction: PropTypes.oneOf(Tooltip.directions),
  noDelay: PropTypes.bool,
  text: PropTypes.string,
  wrap: PropTypes.bool
}
