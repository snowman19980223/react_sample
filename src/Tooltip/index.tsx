import React from 'react'
import {useSSRSafeId} from '@react-aria/ssr'
import type {AnchorPosition, AnchorSide, AnchorAlignment} from '@primer/behaviors'
import Box from '../Box'
import {useAnchoredPosition, useProvidedRefOrCreate} from '../hooks'
import {SxProp, merge, BetterSystemStyleObject} from '../sx'

type TooltipDirection = 'nw' | 'n' | 'ne' | 'e' | 'se' | 's' | 'sw' | 'w'
type TooltipAlign = 'left' | 'right'

export type TooltipProps = {
  /** The text content of the tooltip. This should be brief and no longer than a sentence.
   *  Marked as optional to support backward compatibility with aria-label. */
  text?: string
  /** @deprecated Use `text` instead */
  'aria-label'?: string
  /** Direction relative to target */
  direction?: TooltipDirection
  /** @deprecated Use `direction` instead. Alignment relative to target. */
  align?: TooltipAlign
  /** Use aria-describedby or aria-labelledby */
  type?: 'description' | 'label'
  /** Tooltip target */
  children: React.ReactElement & {ref?: React.RefObject<HTMLElement>}
  /** When set to true, tooltip appears without any delay */
  noDelay?: boolean
  /** @deprecated Always set to true now. */
  wrap?: boolean
} & SxProp

// map tooltip direction to anchoredPosition props
const directionToPosition: Record<TooltipDirection, {side: AnchorSide; align: AnchorAlignment}> = {
  nw: {side: 'outside-top', align: 'start'},
  n: {side: 'outside-top', align: 'center'},
  ne: {side: 'outside-top', align: 'end'},
  e: {side: 'outside-right', align: 'center'},
  se: {side: 'outside-bottom', align: 'end'},
  s: {side: 'outside-bottom', align: 'center'},
  sw: {side: 'outside-bottom', align: 'start'},
  w: {side: 'outside-left', align: 'center'}
}

// map align to AnchorAlignment
const alignToAnchorAlignment: Record<TooltipAlign, AnchorAlignment> = {left: 'start', right: 'end'}

export const TooltipContext = React.createContext<{tooltipId?: string}>({})

export const Tooltip: React.FC<TooltipProps> = ({
  text,
  children,
  direction = 'n',
  align,
  type = 'description',
  noDelay = false,
  sx = {},
  ...props
}) => {
  const tooltipId = useSSRSafeId()

  const childRef = children.ref
  const anchorElementRef = useProvidedRefOrCreate(childRef)
  const tooltipRef = React.useRef<HTMLDivElement>(null)

  const child = React.cloneElement(children, {
    ref: anchorElementRef,
    [type === 'description' ? 'aria-describedby' : 'aria-labelledby']: tooltipId
  })

  const {position} = useAnchoredPosition({
    side: directionToPosition[direction].side,
    // support both algin and direction for backward compatibility
    align: align ? alignToAnchorAlignment[align] : directionToPosition[direction].align,
    floatingElementRef: tooltipRef,
    anchorElementRef
  })

  const tooltipText = text || props['aria-label']

  return (
    <Box
      as="span"
      sx={{
        ':hover, :focus-within': {'[data-component=tooltip]': {visibility: 'visible', opacity: 1}},
        lineHeight: 1 // we don't want the span wrapper to cause misalignment
      }}
    >
      <TooltipContext.Provider value={{tooltipId}}>{child}</TooltipContext.Provider>
      <FloatingTooltip
        id={tooltipId}
        text={tooltipText}
        noDelay={noDelay}
        ref={tooltipRef}
        position={position}
        sx={sx}
      />
    </Box>
  )
}

const FloatingTooltip = React.forwardRef<
  HTMLDivElement,
  Pick<TooltipProps, 'text' | 'noDelay' | 'sx'> & {id: string; position?: AnchorPosition}
>(({id, text, noDelay, position, sx = {}}, ref) => {
  const styles: BetterSystemStyleObject = {
    visibility: 'hidden',
    opacity: 0,
    transition: 'opacity 100ms ease-in',
    transitionDelay: noDelay ? '0ms' : '400ms',

    backgroundColor: 'neutral.emphasisPlus',
    color: 'fg.onEmphasis',
    borderRadius: 1,
    fontSize: 0,
    paddingY: 1,
    paddingX: 2,
    width: 'fit-content',
    maxWidth: '250px',
    textAlign: 'center',
    position: 'absolute',
    zIndex: 2,
    top: position?.top,
    left: position?.left,

    ':before': {
      content: '""',
      width: 0,
      height: 0,
      border: '5px solid transparent',
      position: 'absolute'
    },

    '&[data-side=outside-top]::before': {
      borderTop: '5px solid',
      borderTopColor: 'neutral.emphasisPlus',
      top: '100%'
    },
    '&[data-side=outside-bottom]::before': {
      borderBottom: '5px solid',
      borderBottomColor: 'neutral.emphasisPlus',
      top: '-10px'
    },
    '&[data-side=outside-left]::before': {
      borderLeft: '5px solid',
      borderLeftColor: 'neutral.emphasisPlus',
      top: 'calc(50% - 5px)',
      left: '100%'
    },
    '&[data-side=outside-right]::before': {
      borderRight: '5px solid',
      borderRightColor: 'neutral.emphasisPlus',
      top: 'calc(50% - 5px)',
      left: '-10px'
    },

    '&[data-align=start][data-side=outside-top]::before, &[data-align=start][data-side=outside-bottom]::before': {
      left: '8px'
    },
    '&[data-align=center][data-side=outside-top]::before, &[data-align=center][data-side=outside-bottom]::before': {
      left: 'calc(50% - 4px)'
    },
    '&[data-align=end][data-side=outside-top]::before, &[data-align=end][data-side=outside-bottom]::before': {
      left: 'calc(100% - 16px)'
    }
  }

  return (
    <Box
      role="tooltip"
      id={id}
      aria-hidden
      data-component="tooltip"
      data-side={position?.anchorSide}
      data-align={position?.anchorAlign}
      ref={ref}
      sx={merge<BetterSystemStyleObject>(styles, sx)}
    >
      {text}
    </Box>
  )
})
