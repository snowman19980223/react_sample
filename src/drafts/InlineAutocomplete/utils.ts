import getCaretCoordinates from '@koddsson/textarea-caret'
import {Children, EventHandler, SyntheticEvent} from 'react'

import {Coordinates, ShowSuggestionsEvent, Suggestion, TextInputCompatibleChild, Trigger} from './types'

const singleWordTriggerTerminators = new Set([' ', '\n'])
const multiWordTriggerTerminators = new Set(['.', '\n'])

const isWhitespace = (char: string) => /\s/.test(char)

/**
 * Calculate whether or not suggestions should be shown based on the given state of the
 * input. If they should be shown, returns the show event.
 */
export const calculateSuggestionsQuery = (
  triggers: Array<Trigger>,
  text: string,
  caretLocation: number
): ShowSuggestionsEvent | null => {
  // Build backwards from the caret location until the most recent trigger character or terminator
  for (
    let i = caretLocation - 1, query = '', potentialTriggers = triggers;
    i >= 0 && potentialTriggers.length > 0;
    i--
  ) {
    const character = text[i]

    if (singleWordTriggerTerminators.has(character)) potentialTriggers = potentialTriggers.filter(t => t.multiWord)
    if (multiWordTriggerTerminators.has(character)) potentialTriggers = potentialTriggers.filter(t => !t.multiWord)

    for (const trigger of potentialTriggers.filter(t => character === t.triggerChar)) {
      // Trigger chars must always be preceded by whitespace or be the first character in the input,
      // and even a multi-word query cannot start with whitespace
      if ((i === 0 || isWhitespace(text[i - 1])) && !isWhitespace(query[0])) return {trigger, query}

      potentialTriggers = potentialTriggers.filter(t => t !== trigger)
    }

    query = character + query
  }

  return null
}

/**
 * Obtain the coordinates (px) of the bottom left of a character in an input, relative to the
 * top-left corner of the input itself.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 * @param adjustForScroll Control whether the returned value is adjusted based on scroll position.
 */
export const getCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number,
  adjustForScroll = true
): Coordinates => {
  if (!input) return {top: 0, left: 0}

  // word-wrap:break-word breaks the getCaretCoordinates calculations (a bug), and word-wrap has
  // no effect on input element anyway
  if (input instanceof HTMLInputElement) input.style.wordWrap = ''

  let coords = getCaretCoordinates(input, index)

  // The library calls parseInt on the computed line-height of the element, failing to account for
  // the possibility of it being 'normal' (another bug). In that case, fall back to a rough guess
  // of 1.2 based on MDN: "Desktop browsers use a default value of roughly 1.2".
  if (isNaN(coords.height)) coords.height = parseInt(getComputedStyle(input).fontSize) * 1.2

  // Sometimes top is negative, incorrectly, because of the wierd line-height calculations around
  // border-box sized single-line inputs.
  coords.top = Math.abs(coords.top)

  // For some single-line inputs, the rightmost character can be accidentally wrapped even with the
  // wordWrap fix above. If this happens, go back to the last usable index
  let adjustedIndex = index
  while (input instanceof HTMLInputElement && coords.top > coords.height) {
    coords = getCaretCoordinates(input, --adjustedIndex)
  }

  const scrollTopOffset = adjustForScroll ? -input.scrollTop : 0
  const scrollLeftOffset = adjustForScroll ? -input.scrollLeft : 0

  return {top: coords.top + coords.height + scrollTopOffset, left: coords.left + scrollLeftOffset}
}

/**
 * Obtain the coordinates of the bottom left of a character in an input relative to the top-left
 * of the page.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 */
export const getAbsoluteCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number
): Coordinates => {
  const {top: relativeTop, left: relativeLeft} = getCharacterCoordinates(input, index, true)
  const {top: viewportOffsetTop, left: viewportOffsetLeft} = input?.getBoundingClientRect() ?? {top: 0, left: 0}

  return {
    top: viewportOffsetTop + relativeTop,
    left: viewportOffsetLeft + relativeLeft
  }
}

export const getSuggestionValue = (suggestion: Suggestion): string =>
  typeof suggestion === 'string' ? suggestion : suggestion.value

export const getSuggestionKey = (suggestion: Suggestion): string =>
  typeof suggestion === 'string' ? suggestion : suggestion.key ?? suggestion.value

/**
 * Replace a section of a string.
 */
export const replaceSlice = (
  original: string,
  [startInclusive, endExclusive]: [number, number],
  replacement: string
) => {
  const before = original.substring(0, startInclusive)
  const after = original.substring(endExclusive)
  return before + replacement + after
}

/**
 * Attempts to assert that the child element is of a supported type. This can't be enforced
 * by the type system so it has to be done as a runtime check. This isn't foolproof - a
 * component that forwards a ref to a correct element but does not forward event handlers
 * will not work. But it's the best we can reasonably do.
 */
export function requireChildrenToBeInput(
  child: React.ReactElement,
  childRef: React.RefObject<HTMLElement>
): TextInputCompatibleChild {
  Children.only(child) // Assert that the child is lonely
  if (
    // There is no way to know what type the underlying child is until it mounts, so this
    // will always pass on first render before failing on the second render
    childRef.current &&
    !(childRef.current instanceof HTMLInputElement) &&
    !(childRef.current instanceof HTMLTextAreaElement)
  ) {
    throw new TypeError(
      `AutocompleteTextarea child must be a component that forwards a ref and props to an <input> or <textarea> element.`
    )
  }
  return child
}

/**
 * Combine several event handlers into one. The last handler in the list is called first
 * and no further handlers will be called if `event.preventDefault()` is called.
 */
export const augmentHandler =
  <E extends SyntheticEvent>(...handlers: Array<EventHandler<E> | undefined>) =>
  (event: E) => {
    for (const handler of [...handlers].reverse()) {
      if (!event.isDefaultPrevented()) handler?.(event)
    }
  }
