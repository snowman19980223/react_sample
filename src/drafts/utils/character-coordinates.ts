import getCaretCoordinates from '@koddsson/textarea-caret'

export type Coordinates = {
  top: number
  left: number
}

/**
 * Obtain the coordinates (px) of the bottom left of a character in an input, relative to the
 * top-left corner of the input itself.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 * @param adjustForScroll Control whether the returned value is adjusted based on scroll position.
 */
export const getCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number,
  adjustForScroll = true
): Coordinates => {
  if (!input) return {top: 0, left: 0}

  // word-wrap:break-word breaks the getCaretCoordinates calculations (a bug), and word-wrap has
  // no effect on input element anyway
  if (input instanceof HTMLInputElement) input.style.wordWrap = ''

  let coords = getCaretCoordinates(input, index)

  // The library calls parseInt on the computed line-height of the element, failing to account for
  // the possibility of it being 'normal' (another bug). In that case, fall back to a rough guess
  // of 1.2 based on MDN: "Desktop browsers use a default value of roughly 1.2".
  if (isNaN(coords.height)) coords.height = parseInt(getComputedStyle(input).fontSize) * 1.2

  // Sometimes top is negative, incorrectly, because of the wierd line-height calculations around
  // border-box sized single-line inputs.
  coords.top = Math.abs(coords.top)

  // For some single-line inputs, the rightmost character can be accidentally wrapped even with the
  // wordWrap fix above. If this happens, go back to the last usable index
  let adjustedIndex = index
  while (input instanceof HTMLInputElement && coords.top > coords.height) {
    coords = getCaretCoordinates(input, --adjustedIndex)
  }

  const scrollTopOffset = adjustForScroll ? -input.scrollTop : 0
  const scrollLeftOffset = adjustForScroll ? -input.scrollLeft : 0

  return {top: coords.top + coords.height + scrollTopOffset, left: coords.left + scrollLeftOffset}
}

/**
 * Obtain the coordinates of the bottom left of a character in an input relative to the top-left
 * of the page.
 * @param input The target input element.
 * @param index The index of the character to calculate for.
 */
export const getAbsoluteCharacterCoordinates = (
  input: HTMLTextAreaElement | HTMLInputElement | null,
  index: number
): Coordinates => {
  const {top: relativeTop, left: relativeLeft} = getCharacterCoordinates(input, index, true)
  const {top: viewportOffsetTop, left: viewportOffsetLeft} = input?.getBoundingClientRect() ?? {top: 0, left: 0}

  return {
    top: viewportOffsetTop + relativeTop,
    left: viewportOffsetLeft + relativeLeft
  }
}
