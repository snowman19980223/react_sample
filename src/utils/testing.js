import renderer from 'react-test-renderer'
import enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

enzyme.configure({adapter: new Adapter()})

/**
 * Render the component (a React.createElement() or JSX expression)
 * into its intermediate object representation with 'type',
 * 'props', and 'children' keys
 *
 * The returned object can be matched with expect().toEqual(), e.g.
 *
 * ```js
 * expect(render(<Foo />)).toEqual(render(<div foo='bar' />))
 * ```
 */
export function render(component) {
  return renderer.create(component).toJSON()
}

/**
 * Get the HTML class names rendered by the component instance
 * as an array.
 *
 * ```js
 * expect(renderClasses(<div className='a b' />))
 *   .toEqual(['a', 'b'])
 * ```
 */
export function renderClasses(component) {
  const {
    props: {className}
  } = render(component)
  return className ? className.trim().split(' ') : []
}

/**
 * console.error() is ugly af in jest; mock it with a noop,
 * and return the mock so that you can restore it:
 *
 * ```js
 * const mock = silenceConsoleError(jest)
 * // do stuff that calls console.error()
 * mock.mockRestore() // restore console.error()
 * ```
 *
 * Note: we need to pass `jest` as a reference here because
 * jest's mocking functions are bound to a module at runtime, and
 * need to be called as methods rather than pure functions
 * exported from the jest-mock module.
 */
export function silenceConsoleError(jest) {
  return jest.spyOn(console, 'error').mockImplementation(jest.fn())
}

export function mount(component) {
  return enzyme.mount(component)
}
