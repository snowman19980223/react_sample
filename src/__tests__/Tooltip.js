import React from 'react'
import Tooltip from '../Tooltip'
import {render, renderClasses} from '../utils/testing'

describe('Tooltip', () => {
  it('renders a <span> with the "tooltipped" class', () => {
    expect(render(<Tooltip />).type).toEqual('span')
    expect(renderClasses(<Tooltip />)).toEqual(['tooltipped', 'tooltipped-n'])
  })

  it('respects the "align" prop', () => {
    expect(rendersClass(<Tooltip align="left" />, 'tooltipped-align-left-2')).toBe(true)
    expect(rendersClass(<Tooltip align="right" />, 'tooltipped-align-right-2')).toBe(true)
  })

  it('respects the "direction" prop', () => {
    for (const direction of Tooltip.directions) {
      expect(renderClasses(<Tooltip direction={direction} />).includes(`tooltipped-${direction}`)).toBe(true)
    }
  })

  it('respects the "noDelay" prop', () => {
    expect(renderClasses(<Tooltip noDelay />).includes(`tooltipped-no-delay`)).toBe(true)
  })

  it('respects the "text" prop', () => {
    expect(render(<Tooltip text="hi" />).props['aria-label']).toEqual('hi')
  })

  it('respects the "wrap" prop', () => {
    expect(renderClasses(<Tooltip wrap />).includes('tooltipped-multiline')).toBe(true)
  })
})
