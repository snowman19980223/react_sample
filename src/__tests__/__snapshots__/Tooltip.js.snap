// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders consistently 1`] = `
.c0 {
  position: relative;
}

.c0::before {
  position: absolute;
  z-index: 1000001;
  display: none;
  width: 0px;
  height: 0px;
  color: #1b1f23;
  pointer-events: none;
  content: '';
  border: 6px solid transparent;
  opacity: 0;
}

.c0::after {
  position: absolute;
  z-index: 1000000;
  display: none;
  padding: 0.5em 0.75em;
  font: normal normal 11px/1.5 -apple-system,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji";
  -webkit-font-smoothing: subpixel-antialiased;
  color: #fff;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-shadow: none;
  text-transform: none;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  word-wrap: break-word;
  white-space: pre;
  pointer-events: none;
  content: attr(aria-label);
  background: #1b1f23;
  border-radius: 3px;
  opacity: 0;
}

.c0:hover::before,
.c0:active::before,
.c0:focus::before,
.c0:hover::after,
.c0:active::after,
.c0:focus::after {
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-animation-name: tooltip-appear;
  animation-name: tooltip-appear;
  -webkit-animation-duration: 0.1s;
  animation-duration: 0.1s;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  -webkit-animation-timing-function: ease-in;
  animation-timing-function: ease-in;
  -webkit-animation-delay: 0.4s;
  animation-delay: 0.4s;
}

.c0.tooltipped-no-delay:hover::before,
.c0.tooltipped-no-delay:active::before,
.c0.tooltipped-no-delay:focus::before,
.c0.tooltipped-no-delay:hover::after,
.c0.tooltipped-no-delay:active::after,
.c0.tooltipped-no-delay:focus::after {
  -webkit-animation-delay: 0s;
  animation-delay: 0s;
}

.c0.tooltipped-multiline:hover::after,
.c0.tooltipped-multiline:active::after,
.c0.tooltipped-multiline:focus::after {
  display: table-cell;
}

.c0.tooltipped-s::after,
.c0.tooltipped-se::after,
.c0.tooltipped-sw::after {
  top: 100%;
  right: 50%;
  margin-top: 6px;
}

.c0.tooltipped-s::before,
.c0.tooltipped-se::before,
.c0.tooltipped-sw::before {
  top: auto;
  right: 50%;
  bottom: -7px;
  margin-right: -6px;
  border-bottom-color: #1b1f23;
}

.c0.tooltipped-se::after {
  right: auto;
  left: 50%;
  margin-left: -16px;
}

.c0.tooltipped-sw::after {
  margin-right: -16px;
}

.c0.tooltipped-n::after,
.c0.tooltipped-ne::after,
.c0.tooltipped-nw::after {
  right: 50%;
  bottom: 100%;
  margin-bottom: 6px;
}

.c0.tooltipped-n::before,
.c0.tooltipped-ne::before,
.c0.tooltipped-nw::before {
  top: -7px;
  right: 50%;
  bottom: auto;
  margin-right: -6px;
  border-top-color: #1b1f23;
}

.c0.tooltipped-ne::after {
  right: auto;
  left: 50%;
  margin-left: -16px;
}

.c0.tooltipped-nw::after {
  margin-right: -16px;
}

.c0.tooltipped-s::after,
.c0.tooltipped-n::after {
  -webkit-transform: translateX(50%);
  -ms-transform: translateX(50%);
  transform: translateX(50%);
}

.c0.tooltipped-w::after {
  right: 100%;
  bottom: 50%;
  margin-right: 6px;
  -webkit-transform: translateY(50%);
  -ms-transform: translateY(50%);
  transform: translateY(50%);
}

.c0.tooltipped-w::before {
  top: 50%;
  bottom: 50%;
  left: -7px;
  margin-top: -6px;
  border-left-color: #1b1f23;
}

.c0.tooltipped-e::after {
  bottom: 50%;
  left: 100%;
  margin-left: 6px;
  -webkit-transform: translateY(50%);
  -ms-transform: translateY(50%);
  transform: translateY(50%);
}

.c0.tooltipped-e::before {
  top: 50%;
  right: -7px;
  bottom: 50%;
  margin-top: -6px;
  border-right-color: #1b1f23;
}

.c0.tooltipped-multiline::after {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  max-width: 250px;
  word-wrap: break-word;
  white-space: pre-line;
  border-collapse: separate;
}

.c0.tooltipped-multiline.tooltipped-s::after,
.c0.tooltipped-multiline.tooltipped-n::after {
  right: auto;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c0.tooltipped-multiline.tooltipped-w::after,
.c0.tooltipped-multiline.tooltipped-e::after {
  right: 100%;
}

.c0.tooltipped-align-right-2::after {
  right: 0;
  margin-right: 0;
}

.c0.tooltipped-align-right-2::before {
  right: 15px;
}

.c0.tooltipped-align-left-2::after {
  left: 0;
  margin-left: 0;
}

.c0.tooltipped-align-left-2::before {
  left: 10px;
}

<span
  className="c0 tooltipped-n"
  theme={
    Object {
      "borderWidths": Array [
        0,
        "1px",
      ],
      "breakpoints": Array [
        "544px",
        "768px",
        "1012px",
        "1280px",
      ],
      "buttons": Object {
        "danger": Object {
          "bg": Object {
            "active": "#be222e",
            "default": "#fafbfc",
            "disabled": "#F3F4F6",
            "hover": "#cb2431",
          },
          "border": Object {
            "active": "#9e1c23",
            "default": "#e1e4e8",
            "hover": "#b31d28",
          },
          "color": Object {
            "active": "#fff",
            "default": "#cb2431",
            "disabled": "rgba(203,36,49, .5)",
            "hover": "#fff",
          },
          "shadow": Object {
            "active": "inset 0px 2px 0px rgba(179, 29, 40, 0.4)",
            "default": "0px 1px 0px rgba(149, 157, 165, 0.1), inset 0px 2px 0px rgba(255, 255, 255, 0.25)",
            "focus": "0 0 0 3px rgba(203, 36, 49, 0.4)",
            "hover": "0px 1px 0px rgba(149, 157, 165, 0.1), inset 0px 2px 0px rgba(255, 255, 255, 0.03)",
          },
        },
        "default": Object {
          "bg": Object {
            "active": "#edeff2",
            "default": "#fafbfc",
            "disabled": "#fafbfc",
            "hover": "#F3F4F6",
          },
          "border": Object {
            "active": "#d1d5da",
            "default": "#e1e4e8",
            "disabled": "#eaecef",
          },
          "color": Object {
            "default": "#24292e",
            "disabled": "#959da5",
          },
          "shadow": Object {
            "active": "inset 0px 2px 0px rgba(149, 157, 165, 0.1)",
            "default": "0px 1px 0px rgba(149, 157, 165, 0.1), inset 0px 2px 0px rgba(255, 255, 255, 0.25)",
            "focus": "0 0 0 3px rgba(3, 102, 214, 0.3)",
            "hover": "0px 1px 0px rgba(209, 213, 218, 0.2), inset 0px 2px 0px rgba(255, 255, 255, 0.1)",
          },
        },
        "outline": Object {
          "bg": Object {
            "active": "#035fc7",
            "default": "#fafbfc",
            "disabled": "#F3F4F6",
            "hover": "#0366d6",
          },
          "border": Object {
            "active": "rgba(4, 66, 137, .5)",
            "default": "#e1e4e8",
            "hover": "#005cc5",
          },
          "color": Object {
            "active": "#fff",
            "default": "#005cc5",
            "disabled": "#959da5",
            "hover": "#fff",
          },
          "shadow": Object {
            "active": "inset 0px 1px 0px rgba(4, 66, 137, 0.2)",
            "default": "0px 1px 0px rgba(149, 157, 165, 0.1), inset 0px 2px 0px rgba(255, 255, 255, 0.25)",
            "focus": "0 0 0 3px rgba(3, 102, 214, 0.3)",
            "hover": "0px 1px 0px rgba(149, 157, 165, 0.1), inset 0px 2px 0px rgba(255, 255, 255, 0.03)",
          },
        },
        "primary": Object {
          "bg": Object {
            "active": "#128031",
            "default": "#159739",
            "disabled": "#94D3A2",
            "focus": "#138934",
            "hover": "#138934",
          },
          "border": Object {
            "active": "#176f2c",
            "default": "#22863a",
            "disabled": "rgba(34, 134, 58, 0.1)",
            "hover": "#176f2c",
          },
          "color": Object {
            "default": "#fff",
            "disabled": "rgba(255, 255, 255, 0.50)",
          },
          "shadow": Object {
            "active": "inset 0px 1px 0px rgba(20, 70, 32, 0.2)",
            "default": " 0px 1px 0px rgba(20, 70, 32, 0.1), inset 0px 2px 0px rgba(255, 255, 255, 0.03)",
            "focus": "0 0 0 3px #94D3A2",
          },
        },
      },
      "colors": Object {
        "accent": "#f66a0a",
        "bg": Object {
          "disabled": "#F3F4F6",
          "gray": "#f6f8fa",
          "grayLight": "#fafbfc",
        },
        "black": "#1b1f23",
        "blackfade15": "rgba(27, 31, 35, 0.15)",
        "blackfade20": "rgba(27, 31, 35, 0.20)",
        "blackfade30": "rgba(27,31,35,0.3)",
        "blackfade35": "rgba(27, 31, 35, 0.35)",
        "blackfade50": "rgba(27, 31, 35, 0.5)",
        "blue": Array [
          "#f1f8ff",
          "#dbedff",
          "#c8e1ff",
          "#79b8ff",
          "#2188ff",
          "#0366d6",
          "#005cc5",
          "#044289",
          "#032f62",
          "#05264c",
        ],
        "bodytext": "#24292e",
        "border": Object {
          "blackFade": "rgba(27,31,35,0.15)",
          "blue": "#0366d6",
          "blueLight": "#c8e1ff",
          "gray": "#e1e4e8",
          "grayDark": "#d1d5da",
          "grayDarker": "#444d56",
          "grayLight": "#eaecef",
          "green": "#34d058",
          "greenLight": "#a2cbac",
          "purple": "#6f42c1",
          "red": "#d73a49",
          "redLight": "#cea0a5",
          "white": "#fff",
          "whiteFade": "rgba(255,255,255,0.15)",
          "yellow": "#d9d0a5",
        },
        "counter": Object {
          "bg": "rgba(27, 31, 35, 0.08)",
        },
        "filterList": Object {
          "hoverBg": "#eaecef",
        },
        "gray": Array [
          "#fafbfc",
          "#f6f8fa",
          "#e1e4e8",
          "#d1d5da",
          "#959da5",
          "#6a737d",
          "#586069",
          "#444d56",
          "#2f363d",
          "#24292e",
        ],
        "green": Array [
          "#f0fff4",
          "#dcffe4",
          "#bef5cb",
          "#85e89d",
          "#34d058",
          "#28a745",
          "#22863a",
          "#176f2c",
          "#165c26",
          "#144620",
        ],
        "orange": Array [
          "#fff8f2",
          "#ffebda",
          "#ffd1ac",
          "#ffab70",
          "#fb8532",
          "#f66a0a",
          "#e36209",
          "#d15704",
          "#c24e00",
          "#a04100",
        ],
        "purple": Array [
          "#f5f0ff",
          "#e6dcfd",
          "#d1bcf9",
          "#b392f0",
          "#8a63d2",
          "#6f42c1",
          "#5a32a3",
          "#4c2889",
          "#3a1d6e",
          "#29134e",
        ],
        "red": Array [
          "#ffeef0",
          "#ffdce0",
          "#fdaeb7",
          "#f97583",
          "#ea4a5a",
          "#d73a49",
          "#cb2431",
          "#b31d28",
          "#9e1c23",
          "#86181d",
        ],
        "state": Object {
          "error": "#d73a49",
          "failure": "#d73a49",
          "pending": "#dbab09",
          "queued": "#dbab09",
          "success": "#28a745",
          "unknown": "#959da5",
        },
        "text": Object {
          "gray": "#586069",
          "grayDark": "#24292e",
          "grayLight": "#6a737d",
          "red": "#cb2431",
        },
        "white": "#fff",
        "whitefade15": "rgba(255, 255, 255, 0.15)",
        "whitefade50": "rgba(255, 255, 255, 0.50)",
        "yellow": Array [
          "#fffdef",
          "#fffbdd",
          "#fff5b1",
          "#ffea7f",
          "#ffdf5d",
          "#ffd33d",
          "#f9c513",
          "#dbab09",
          "#b08800",
          "#735c0f",
        ],
      },
      "fontSizes": Array [
        "12px",
        "14px",
        "16px",
        "20px",
        "24px",
        "32px",
        "40px",
        "48px",
      ],
      "fontWeights": Object {
        "bold": 600,
        "light": 300,
        "normal": 400,
        "semibold": 500,
      },
      "fonts": Object {
        "mono": "SFMono-Regular, Consolas, \\"Liberation Mono\\", Menlo, Courier, monospace",
        "normal": "-apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\"",
      },
      "lineHeights": Object {
        "condensed": 1.25,
        "condensedUltra": 1,
        "default": 1.5,
      },
      "pagination": Object {
        "borderRadius": "3px",
        "colors": Object {
          "disabled": Object {
            "bg": "#fafbfc",
            "border": "#e1e4e8",
            "fg": "#d1d5da",
          },
          "hover": Object {
            "bg": "#f6f8fa",
            "border": "#e1e4e8",
            "fg": "#0366d6",
          },
          "normal": Object {
            "bg": "#fff",
            "border": "#e1e4e8",
            "fg": "#0366d6",
          },
          "selected": Object {
            "bg": "#0366d6",
            "border": "#0366d6",
            "fg": "#fff",
          },
        },
        "fontSize": "13px",
        "fontWeight": 600,
      },
      "popovers": Object {
        "colors": Object {
          "caret": "rgba(27, 31, 35, 0.15)",
        },
      },
      "radii": Array [
        "0",
        "3px",
        "6px",
        "100px",
      ],
      "shadows": Object {
        "extra-large": "0 12px 48px rgba(149, 157, 165, 0.3)",
        "formControl": "inset 0px 2px 0px rgba(225, 228, 232, 0.2)",
        "formControlDisabled": "inset 0px 2px 0px rgba(220, 227, 237, 0.3)",
        "formControlFocus": "rgba(3, 102, 214, 0.3) 0px 0px 0px 0.2em",
        "large": "0 8px 24px rgba(149, 157, 165, 0.2)",
        "medium": "0 3px 6px rgba(149, 157, 165, 0.15)",
        "primaryActiveShadow": "inset 0px 1px 0px rgba(20, 70, 32, 0.2)",
        "primaryShadow": "0px 1px 0px rgba(20, 70, 32, 0.1), inset 0px 2px 0px rgba(255, 255, 255, 0.03)",
        "small": "0 1px 0 rgba(149, 157, 165, 0.1)",
      },
      "sizes": Object {
        "large": "1012px",
        "medium": "768px",
        "small": "544px",
        "xlarge": "1280px",
      },
      "space": Array [
        "0",
        "4px",
        "8px",
        "16px",
        "24px",
        "32px",
        "40px",
        "48px",
        "64px",
        "80px",
        "96px",
        "112px",
        "128px",
      ],
    }
  }
/>
`;
